### PR TITLE
feat: support builtin kernel mods

### LINKED ISSUE #1025: `kernel.loadedmodule` doesn't work on all OS
**What would you like to be added**:
Currently, `kernel.loadedmodule` feature only check for the content of `/proc/modules` for modules compiled as module and not built in tree module. Maybe there's a way to check for built-in kernel that is in `/lib/modules/<kernel-version>/modules.builtin`.
I think this can be done by either improve the function `getLoadedModules` or allow the user to choose what kind of test to get the information.

**Why is this needed**:

Currently [Talos Linux](https://www.talos.dev/) will just silently not creating `nodeFeatureRule` when I put `matchFeatures` that check for `kernel.loadedmodule` even if the module is loaded.

### PR DESCRIPTION
**Pull request by AhmedGrati**
Resolves #1025.
The solution parsed the `/lib/modules/<KERNEL_VERSION>/modules.builtin` and combined it with the `/proc/modules` file to get most kernel modules.

### CODE REVIEW

File: source/kernel/modules.go [hunk 1]
```diff
@@ -25,20 +26,39 @@ import (
...
-		}
-		// append loaded module
-		loadedMods = append(loadedMods, strings.Fields(line)[0])
-	}
+  kVersion, err := getVersion()
+  var filesToLoad []string
+  if err != nil {
+    filesToLoad = []string{kmodProcfsPath}
+  } else {
+    kBuiltinModPath := "/lib/modules/" + kVersion + "/modules.builtin"
```
Comment (marquiz) [inline source/kernel/modules.go:34]: We need to specify new `LibDir` in `pkg/utils/hostpath/hostpath.go` and use it here.

And, add corresponding mounts to all deployment files (kustomize and Helm)

File: source/kernel/modules.go [hunk 2]
```diff
@@ -25,20 +26,39 @@ import (
...
+    filesToLoad = []string{kmodProcfsPath, kBuiltinModPath}
+  }
+  loadedMods, err := loadModules(filesToLoad)
+  if err != nil {
+    return nil, errors.New("error while loading modules")
+  }
 	return loadedMods, nil
 }
+
+func loadModules(kModFsPath []string) ([] string, error) {
```
Comment (marquiz) [inline source/kernel/modules.go:44]: The format of `/proc/modules` and  `modules.builtin` is different so you need to have separate parser functions for them

**Review by marquiz — COMMENTED** [review]
Thanks @AhmedGrati for working on this.

I'm not entirely sure we want to change the existing `kernel.loadedmodule` feature. If not, we want to document that it lists loaded dynamic kernel modules and omits builtin modules.

Instead this PR could add a new feature, say `kernel.enabledmodule` which would combine loaded dynamic modules and the builtins.

OTOH, if `kernel.loadedmodule` is seen totally useless/redundant maybe we just change that (even if the naming is a bit off).

Any comments @zvonkok @mythi @ArangoGutierrez?

**Comment by ArangoGutierrez** [comment]
PING @AhmedGrati

**Comment by AhmedGrati** [comment]
@ArangoGutierrez Just discussed with @marquiz today about what should be done, since we were waiting for inputs from others. I'll start working on it.

**Comment by ArangoGutierrez** [comment]
Still a WIP
`make test e2e-test` passes

File: source/kernel/modules.go [hunk 3]
```diff
@@ -42,3 +44,35 @@ func getLoadedModules() ([]string, error) {
...
+		// append loaded module
+		loadedMods = append(loadedMods, getBuiltinModName(line))
+	}
+	return loadedMods, nil
+}
+
+func getBuiltinModName(modPath string) string {
+	splittedModPath := strings.Split(modPath, "/")
+	splittedModFileName := strings.Split(splittedModPath[len(splittedModPath)-1], ".")
+	return strings.Join(splittedModFileName[:len(splittedModFileName)-1], ".")
```
Comment (marquiz) [inline source/kernel/modules.go:77]: You could just use strings.TrimSuffix() to drop the `.ko` suffix

File: source/kernel/kernel.go [hunk 4]
```diff
@@ -141,6 +142,12 @@ func (s *kernelSource) Discover() error {
 		s.features.Flags[LoadedModuleFeature] = nfdv1alpha1.NewFlagFeatures(kmods...)
 	}
 
+	if builtinMods, err := getBuiltinModules(); err != nil {
+		klog.Errorf("failed to get builtin kernel modules: %v", err)
+	} else {
+		s.features.Flags[EnabledModuleFeature] = nfdv1alpha1.NewFlagFeatures(builtinMods...)
```
Comment (marquiz) [inline source/kernel/kernel.go:148]: I think this feature should combine loadedmodules and builtinmodules, as discussed in #1025

File: source/kernel/modules.go [hunk 5]
```diff
@@ -42,3 +44,35 @@ func getLoadedModules() ([]string, error) {
...
+	}
+
+	kBuiltinModPath := hostpath.LibDir.Path("modules/" + kVersion + "/modules.builtin")
+	out, err := os.ReadFile(kBuiltinModPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file %s: %s", kBuiltinModPath, err.Error())
+	}
+
+	lines := strings.Split(string(out), "\n")
+	loadedMods := make([]string, len(lines))
```
Comment (marquiz) [inline source/kernel/modules.go:61]: nit: they're not "loaded" (like dynamically) so I'd rename this just `mods` or then `builtinMods`

**Review by marquiz — COMMENTED** [review]
Thanks @AhmedGrati, looks really good. I wanna test this myself 😅

We also need to document the new feature in the table of available features in `docs/usage/customization-guide.md`

File: source/kernel/modules.go [hunk 6]
```diff
@@ -42,3 +45,29 @@ func getLoadedModules() ([]string, error) {
...
+		return nil, fmt.Errorf("failed to read file %s: %s", kBuiltinModPath, err.Error())
+	}
+
+	lines := strings.Split(string(out), "\n")
+	builtinMods := make([]string, len(lines))
+	for _, line := range lines {
+		// skip empty lines
+		if len(line) == 0 {
+			continue
+		}
```
Comment (marquiz) [inline source/kernel/modules.go:67]: One suggestion: instead of checking for empty lines it would prolly be more robust to first trim all white space and then verify that the string ends with .ko (all the modules do AFAIU)

```suggestion
		line = strings.TrimSpace(line)
		if !strings.HasSuffix(line, ".ko") {
			continue
		}
```

**Review by marquiz — COMMENTED** [review]
Thx! Just one comment but basically looks ready for me (after that) 👍

Would be nice to get some other eyes on this, too.
ping @budimanjojo @mythi @zvonkok, any comments?

File: deployment/helm/node-feature-discovery/templates/worker.yaml [hunk 7]
```diff
@@ -99,6 +99,9 @@ spec:
         - name: host-usr-lib
           hostPath:
             path: "/usr/lib"
+        - name: host-lib
```
Comment (marquiz) [inline deployment/helm/node-feature-discovery/templates/worker.yaml:102]: Hey, you need to add  `host-lib` to the mounts of the container above, too

File: deployment/components/common/worker-mounts.yaml [hunk 8]
```diff
@@ -13,6 +13,9 @@
   - name: host-usr-lib
     hostPath:
       path: "/usr/lib"
+  - name: host-lib
```
Comment (marquiz) [inline deployment/components/common/worker-mounts.yaml:16]: Hey, you need to add  `host-lib` to the mounts of the container above, too

**Comment by budimanjojo** [comment]
Looks good to me. This means if I use `kernel.enabledModules` it will detect both the old and new implementations while `kernel.loadedModules` will still behave like the old implementation?

File: deployment/components/common/worker-mounts-combined.yaml [hunk 9]
```diff
@@ -38,6 +41,9 @@
   - name: host-usr-lib
     mountPath: "/host-usr/lib"
     readOnly: true
+  - name: host-lib
+    hostPath:
+      path: "/lib"
```
Comment (marquiz) [inline deployment/components/common/worker-mounts-combined.yaml:46]: The field in volumeMounts is `mountPath`. Also, you need to mount it with the `host-` prefix


```suggestion
    mountPath: "/host-lib"
    readOnly: true
```

...and ditto volumemounts you added into other files

**Comment by marquiz** [comment]
> This means if I use `kernel.enabledModules` it will detect both the old and new implementations while `kernel.loadedModules` will still behave like the old implementation?

Yup, that's the way it works

File: deployment/helm/node-feature-discovery/templates/worker.yaml [hunk 10]
```diff
@@ -67,6 +67,9 @@ spec:
         - name: host-usr-lib
           mountPath: "/host-usr/lib"
           readOnly: true
+        - name: host-lib
+          mountPath: "/lib"
```
Comment (marquiz) [inline deployment/helm/node-feature-discovery/templates/worker.yaml:71]: here, too

File: source/kernel/modules.go [hunk 11]
```diff
@@ -42,3 +45,30 @@ func getLoadedModules() ([]string, error) {
...
+	}
+
+	kBuiltinModPath := hostpath.LibDir.Path("modules/" + kVersion + "/modules.builtin")
+	out, err := os.ReadFile(kBuiltinModPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file %s: %s", kBuiltinModPath, err.Error())
+	}
+
+	lines := strings.Split(string(out), "\n")
+	builtinMods := make([]string, len(lines))
```
Comment (marquiz) [inline source/kernel/modules.go:62]: You need to set the lenght of the array to zero, otherwise it will have no use here (it will have `len(lines)` number of empty strings on which you will append the new values, making the final slice 2 x `len(lines)` long)

```suggestion
	builtinMods := make([]string, 0, len(lines))
```

**Review by marquiz — COMMENTED** [review]
One more issue I spotted.

Also, I'm again nitting about the commit message 😇 Could you add some flesh there, describing what is being added (and possibly also why)

File: docs/usage/customization-guide.md [hunk 12]
```diff
@@ -620,6 +620,7 @@ The following features are available for matching:
 | **`kernel.config`** | attribute |          |            | Kernel configuration options
 |                  |              | **`<config-flag>`** | string | Value of the kconfig option
 | **`kernel.loadedmodule`** | flag |         |            | Loaded kernel modules
+| **`kernel.enabledmodule`** | flag |        |            | Builtin and dynamic kernel modules
```
Comment (mythi) [inline docs/usage/customization-guide.md:623]: this documentation does not explain it very well that `enabledmodule` is a superset of `loadedmodule` and what "buildin modules" are

Comment (marquiz) [reply docs/usage/customization-guide.md:623]: Yeah, now that you pointed this out, I agree. Any suggestions how to improve this @mythi? _All enabled kernel modules, i.e. union of loaded dynamic modules and modules built into the kernel_

Comment (marquiz) [reply docs/usage/customization-guide.md:623]: We could also clarify `kernel.loadedmodule` above i.e. with _Loaded dynamic kernel modules_(?)

File: source/kernel/modules.go (hunk 11 above)
Comment (AhmedGrati) [reply source/kernel/modules.go:62]: @marquiz should we change it also for the `loadedMods := make([]string, len(lines))` (It's in the function above)?

Comment (AhmedGrati) [reply source/kernel/modules.go]: No worries, I'll include it here.

**Comment by marquiz** [comment]
ping @mythi

File: docs/usage/customization-guide.md [hunk 13]
```diff
@@ -619,7 +619,8 @@ The following features are available for matching:
 | | |          **`nx_gzip`**                 | bool       | Nest Accelerator GZIP support is enabled
 | **`kernel.config`** | attribute |          |            | Kernel configuration options
 |                  |              | **`<config-flag>`** | string | Value of the kconfig option
-| **`kernel.loadedmodule`** | flag |         |            | Loaded kernel modules
+| **`kernel.loadedmodule`** | flag |         |            | Loaded dynamic kernel modules
+| **`kernel.enabledmodule`** | flag |        |            | All enabled kernel modules, i.e. union of loaded dynamic modules and modules built into the kernel
```
Comment (mythi) [inline docs/usage/customization-guide.md:623]:
```suggestion
| **`kernel.loadedmodule`** | flag |         |            | Kernel modules loaded on the node as reported by `/proc/modules`
| **`kernel.enabledmodule`** | flag |        |            | Kernel modules loaded on the node and available as built-ins as reported by `modules.builtin`. Does not include any out-of-tree modules that are not loaded.
```

**Comment by mythi** [comment]
> @mythi WDYT about the new wording(s) in the docs?

I added a suggestion but I'm also OK with how it is currently.

File: docs/usage/customization-guide.md (hunk 13 above)
Comment (marquiz) [reply docs/usage/customization-guide.md:623]: Out-of-tree modules are no special in this regard. Also, we report **loaded** and **enabled-by-builtin** modules so it should be kinda evident that we don't list any **disabled** modules so I'd suggest just to drop this _Does not include any out-of-tree modules that are not loaded._ Or then reword it to _Does not include any modules that are not loaded or built into the kernel._ which sounds a bit dumb after the first sentence

Comment (mythi) [reply docs/usage/customization-guide.md:623]: > I'd suggest just to drop this

works for me

Comment (marquiz) [reply docs/usage/customization-guide.md:623]: K, so @mythi's suggestion would become

```suggestion
| **`kernel.loadedmodule`** | flag |         |            | Kernel modules loaded on the node as reported by `/proc/modules`
| **`kernel.enabledmodule`** | flag |        |            | Kernel modules loaded on the node and available as built-ins as reported by `modules.builtin`
```

I think this is good.

@AhmedGrati could I ask for this one more change? 😇

Comment (AhmedGrati) [reply docs/usage/customization-guide.md:623]: Sure, will do it 👍

**Review by marquiz — APPROVED** [review]
Thanks @AhmedGrati!

**Review by mythi — APPROVED** [review]

**Comment by marquiz** [comment]
CI is failing, both pre-submit and e2e 😭 Let's fix those first
/hold

**Comment by marquiz** [comment]
e2e is green again
/unhold
